### PR TITLE
Migrate CF CLI Bosh release to GitHub Action workflow

### PR DESCRIPTION
**Pull request by a-b**
- Setup devcontainer-based local environment using cached base image
- Use nekto/act to iterate on the pipeline locally
- Extract shepherd leasing option into the separate remote action
- Implenent machanism to sync GHA variables and secrets between repositories

### CODE REVIEW

File: config/final.yml
Comment (a-b) [inline config/final.yml:1]: This will be reverted on after final check

**Review by joaopapereira — CHANGES_REQUESTED** [review]
In general, it looks like a pretty good improvement from before.

As a personal note, this feels like a lot of bash code all around, and when it gets that big, I prefer to use a different programming language with better features instead of relying on bash.

**Review by gururajsh — CHANGES_REQUESTED** [review]

File: .devcontainer/Dockerfile
Comment (a-b) [reply .devcontainer/Dockerfile:1]: 👍 It should not. We still need to figure out how to pass parameters into the docker file. If you have ideas, please let me know.

File: .github/workflows/README.md
Comment (a-b) [reply .github/workflows/README.md:8]: This pipeline is located on the public concourse instance and should be visible to users authenticated via GitHub. However, this instance is deprecated, so we need to remove it anyway.

File: .github/workflows/create-bosh-release.yml
Comment (a-b) [reply .github/workflows/create-bosh-release.yml:20]: Our deployment plan is to run the first pass after the merge in parallel using the shadow branch and S3 bucket suffixed with `-dev` to ensure correctness. IMO that's where we can validate this.

File: .github/workflows/create-bosh-release.yml
Comment (a-b) [reply .github/workflows/create-bosh-release.yml:13]: We're using this for demo / local dev purposes, but it will eventually be removed.

File: .github/workflows/create-bosh-release.yml
Comment (a-b) [reply .github/workflows/create-bosh-release.yml:87]: We will not remove it yet; we will do so during phase 2 of our [deployment plan](https://github.com/cloudfoundry/bosh-package-cf-cli-release/pull/19#discussion_r1635542562).

File: .github/workflows/create-bosh-release.yml
Comment (a-b) [reply .github/workflows/create-bosh-release.yml:88]: `--force` and `-dev` will be removed immediately after the demonstration.

File: .github/workflows/create-bosh-release.yml
Comment (a-b) [reply .github/workflows/create-bosh-release.yml:94]: We have created https://github.com/pivotal-cf/gha-shepherd and are currently working on making it public (bureaucracy). In the meantime, we're using this repository for demonstration purposes.

File: .github/workflows/create-bosh-release.yml
Comment (a-b) [reply .github/workflows/create-bosh-release.yml:124]: In this project, we're actively using [devconainers](https://containers.dev/) and https://nektosact.com/, https://github.com/nektos/act . 

There were [some issues](https://github.com/nektos/act/issues/2351) with `—uses: actions/upload-artifact@v4` and `—uses: actions/download-artifact@v4`. Act [emulates the assets server](https://github.com/a-b/cf-cli-release/blob/b6509ceae25846946d41593e7253ef3083bd0332/Makefile#L58), and the authors of both projects collaborate tightly. 

The problem was resolved by either deleting the file before the download step, downgrading to `v3`, or removing the local assets server folder. Both versions work perfectly in the cloud.

In short, this TODO is now obsolete.

File: .github/workflows/create-bosh-release.yml
Comment (a-b) [reply .github/workflows/create-bosh-release.yml:170]: We install [these dependencies in our base image](https://github.com/a-b/cf-cli-release/blob/b6509ceae25846946d41593e7253ef3083bd0332/ci/Dockerfile#L70-L91), and some are duplicated. I like the method via GitHub releases, but I'm not sure if keeping the latest release is a good idea for reproducibility. This needs to be consolidated.

File: .github/workflows/create-bosh-release.yml
Comment (a-b) [reply .github/workflows/create-bosh-release.yml:219]: I love your suggestion. Unfortunately, I didn't have time to try to [pass this input](https://nektosact.com/usage/index.html#pass-inputs-to-manually-triggered-workflows) via the `act` in the local environment [here](https://github.com/a-b/cf-cli-release/blob/b6509ceae25846946d41593e7253ef3083bd0332/Makefile#L51-L59) yet.

File: .github/workflows/create-bosh-release.yml
Comment (a-b) [reply .github/workflows/create-bosh-release.yml:221]: Thanks for finding this one!

File: .github/workflows/create-bosh-release.yml
Comment (a-b) [reply .github/workflows/create-bosh-release.yml:285]: We need to consider how to adapt this for local use within our development cycle. We will remove it before the merge, but not right now.

File: .github/workflows/create-bosh-release.yml
Comment (a-b) [reply .github/workflows/create-bosh-release.yml:311]: https://github.com/cloudfoundry/bosh-package-cf-cli-release/pull/19#discussion_r1635542562

File: .github/workflows/create-bosh-release.yml
Comment (a-b) [reply .github/workflows/create-bosh-release.yml:331]: This is something I hope to discuss as a group.

### Challenge
There are multiple consumers of this release:
1. TAS releng are using [GitHub releases](https://github.com/cloudfoundry/bosh-package-cf-cli-release/releases)
1. bosh.io workers are using [filesystem](https://github.com/cloudfoundry/bosh-package-cf-cli-release/tree/main/releases/cf-cli)
1. I suspect someone may be downloading archives with releases directly from S3


Our proposed evolution of the current implementation involves the creation of a draft release in the release process. All commits and changes to the blobstore will be pushed after we promote the draft in an atomic way.

File: .github/workflows/lint.yml
Comment (a-b) [reply .github/workflows/lint.yml:8]: Yes, this should be aligned with [this](https://github.com/a-b/cf-cli-release/blob/41390d344af342648329752f3dd6a051ba960bb0/.github/workflows/create-bosh-release.yml#L29)

File: Makefile
Comment (a-b) [reply Makefile:59]: This is the main entry point for iterating on the pipeline locally. What motivates you to remove this? Are you suggesting to move this [here](https://github.com/a-b/cf-cli-release/blob/41390d344af342648329752f3dd6a051ba960bb0/Makefile#L28-L32) ?

File: ci/Dockerfile
Comment (a-b) [reply ci/Dockerfile:78]: https://github.com/cloudfoundry/bosh-package-cf-cli-release/pull/19#discussion_r1635561043

File: ci/scripts/create-bosh-release-candidate.sh
Comment (a-b) [reply ci/scripts/create-bosh-release-candidate.sh:5]: cc @chinigo

File: ci/Dockerfile
Comment (a-b) [reply ci/Dockerfile:11]: cc @chinigo

File: config/final.yml
Comment (a-b) [reply config/final.yml:6]: https://github.com/cloudfoundry/bosh-package-cf-cli-release/pull/19#discussion_r1635542562

File: .github/workflows/create-bosh-release.yml
Comment (a-b) [reply .github/workflows/create-bosh-release.yml:87]: I'd like to discuss our deployment plan as a group. To mitigate some risks, I propose two phases: first, we start with a shadow strategy (branch and S3 bucket are not in production) to check correctness, and then we follow with an update to the production buckets right after.

File: .devcontainer/Dockerfile
Comment (joaopapereira) [reply .devcontainer/Dockerfile:1]: here is an example https://www.jeffgeerling.com/blog/2017/use-arg-dockerfile-dynamic-image-specification and https://github.com/microsoft/vscode-remote-release/issues/3545

File: Makefile
Comment (joaopapereira) [reply Makefile:59]: if it is commented, it will not run, so what is the point of being here? or am I missing something?

File: .devcontainer/Dockerfile
Comment (a-b) [reply .devcontainer/Dockerfile:1]: Thanks - this is helpful!
cc @chinigo for visibility

File: .github/workflows/create-bosh-release.yml
Comment (a-b) [reply .github/workflows/create-bosh-release.yml:124]: This step is used to cache lease json. It is useful for troubleshooting.

File: .github/workflows/create-bosh-release.yml
Comment (a-b) [reply .github/workflows/create-bosh-release.yml:311]: https://github.com/cloudfoundry/bosh-package-cf-cli-release/pull/19#discussion_r1635595572

File: .github/workflows/create-bosh-release.yml
Comment (a-b) [reply .github/workflows/create-bosh-release.yml:312]: This will be removed before the final merge. It is used for demoing this process.

File: .github/workflows/create-bosh-release.yml
Comment (a-b) [reply .github/workflows/create-bosh-release.yml:350]: I'm fine with removing this step, but I'm also okay with leaving it as there are a couple of safeguards in place:
1. Only repository approvers can rerun the job with the debug checkbox.
image

1. The output of these runs is only visible to the core team. Below is an example of a run from the private browser window.
image

File: Makefile
Comment (a-b) [reply Makefile:49]: This is a feature allowing retrigger runs on changes in the filesystem via `entr` utility https://eradman.com/entrproject/

File: ci/Dockerfile
Comment (a-b) [reply ci/Dockerfile:75]: Do you think we should hardcode `bosh`, `bbl`, and other tool versions or keep them to the latest release?  We're storing images for all builds, so that should satisfy the build reproducibility requirement.

File: config/final.yml
Comment (a-b) [reply config/final.yml:6]: https://github.com/cloudfoundry/bosh-package-cf-cli-release/pull/19#discussion_r1635595572

File: .github/workflows/create-bosh-release.yml
Comment (gururajsh) [reply .github/workflows/create-bosh-release.yml:87]: This was already tested in dev branch  with dev buckets correct?

File: .devcontainer/Dockerfile
Comment (a-b) [reply .devcontainer/Dockerfile:1]: I propose we temporarily switch it to `FROM ghcr.io/cloudfoundry/bosh-package-cf-cli-release/cf-cli-release-ci:main` devcontainers should work after our first `Ensure CI image` run and follow up with a proper portability fix.

File: Makefile
Comment (a-b) [reply Makefile:59]: The commented part is only autorunner `make run` works as expected.

File: .github/workflows/create-bosh-release.yml
Comment (a-b) [reply .github/workflows/create-bosh-release.yml:86]: Correct. Considering the involved risks, we still want to do our first pass to the dev bucket, followed by a change to production.

**Pull request merged**
